physionet-django/physionet/gcp.py

```diff
@@ -0,0 +1,179 @@
+
+    @property
+    def dir_blob(self):
+        return self.bucket.blob(self.dir_key)
+
+    def batch(self):
+        return self.client.batch()
+
+    def create_bucket(self, location=None):
+        bucket = self.client.bucket(self.bucket_name)
+        bucket.location = location or settings.GCP_BUCKET_LOCATION
+        bucket.iam_configuration.uniform_bucket_level_access_enabled = True
+        self.client.create_bucket(bucket)
+
+    def create_bucket_if_needed(self):
+        try:
+            self.create_bucket()
+        except Conflict:
+            pass
+
+    def put(self, data):
+        self.blob.upload_from_string(data)
+
+    def put_fileobj(self, file):
+        self.blob.upload_from_file(file)
+
+    def mkdir(self):
+        self.dir_blob.upload_from_string('')
+
+    def exists(self):
+        return self.file_exists() or self.dir_exists()
+
+    def file_exists(self):
+        return self.blob.exists()
+
+    def dir_exists(self):
+        iterator = self.client.list_blobs(self.bucket_name, prefix=self.dir_key, max_results=1)
+        return len(list(iterator)) > 0
+
+    def dir_size(self):
+        iterator = self.client.list_blobs(self.bucket_name, prefix=self.dir_key)
+        return sum([obj.size for obj in iterator])
+
+    def open(self, mode='rb'):
+        storage = GoogleCloudStorage(bucket_name=self.bucket_name)
+        return storage.open(self.key, mode=mode)
+
+    def list_dir(self):
+        iterator = self.client.list_blobs(self.bucket_name, prefix=self.dir_key, delimiter='/')
+        blobs = list(iterator)
+        prefixes = iterator.prefixes
+
+        files = []
+        dirs = []
+
+        for blob in blobs:
+            name = blob.name.replace(self.dir_key, '', 1)
+            if name != '':
+                size = readable_size(blob.size)
+                modified = blob.updated.strftime("%Y-%m-%d")
+                files.append(FileInfo(name, size, modified))
+
+        for prefix in prefixes:
+            dirs.append(DirectoryInfo(prefix.replace(self.dir_key, '', 1)[:-1]))
+
+        files.sort()
+        dirs.sort()
+
+        return files, dirs
+
+    def url(self):
+        storage = GoogleCloudStorage(bucket_name=self.bucket_name)
+        return storage.url(self.key)
+
+    def rm(self):
+        try:
+            self.rm_file()
+        except NotFound:
+            pass
+
+        self.rm_dir()
+
+    def rm_file(self):
+        self.blob.delete()
+
+    def rm_dir(self):
+        blobs = list(self.client.list_blobs(self.bucket_name, prefix=self.dir_key))
+        self.bucket.delete_blobs(blobs=blobs)
+
+    def cp(self, other):
+        try:
+            self.cp_file(other)
+        except NotFound:
+            pass
+
+        self.cp_dir(other)
+
+    def cp_file(self, other):
+        self.bucket.copy_blob(self.blob, other.bucket, new_name=other.key)
+
+    def cp_dir(self, other, ignored_files=[]):
+        ignored_files = [os.path.join(self.dir_key, f) for f in ignored_files]
```

Comment:
It's considered bad practice to have mutable default argument, usually you would do something like:
```suggestion
    def cp_dir(self, other, ignored_files=None):
        ignored_files = ignored_files or []
        ignored_files = [os.path.join(self.dir_key, f) for f in ignored_files]
```